core/src/main/java/org/apache/iceberg/hadoop/HadoopTableOperations.java

```diff
@@ -149,26 +183,71 @@ public void commit(TableMetadata base, TableMetadata metadata) {
     String codecName =
         metadata.property(
             TableProperties.METADATA_COMPRESSION, TableProperties.METADATA_COMPRESSION_DEFAULT);
+    // TODO:This is not compatible with the scenario where the user modifies the metadata file
+    // compression codec arbitrarily.
+    // We can inform the user about this bug first, and fix it later.(Do not modify the compressed
+    // format after the table is created.)
     TableMetadataParser.Codec codec = TableMetadataParser.Codec.fromName(codecName);
     String fileExtension = TableMetadataParser.getFileExtension(codec);
-    Path tempMetadataFile = metadataPath(UUID.randomUUID().toString() + fileExtension);
+    Path tempMetadataFile = metadataPath(UUID.randomUUID() + fileExtension);
     TableMetadataParser.write(metadata, io().newOutputFile(tempMetadataFile.toString()));
 
     int nextVersion = (current.first() != null ? current.first() : 0) + 1;
     Path finalMetadataFile = metadataFilePath(nextVersion, codec);
     FileSystem fs = getFileSystem(tempMetadataFile, conf);
-
-    // this rename operation is the atomic commit operation
-    renameToFinal(fs, tempMetadataFile, finalMetadataFile, nextVersion);
-
-    LOG.info("Committed a new metadata file {}", finalMetadataFile);
-
-    // update the best-effort version pointer
-    writeVersionHint(nextVersion);
-
-    deleteRemovedMetadataFiles(base, metadata);
-
-    this.shouldRefresh = true;
+    boolean versionCommitSuccess = false;
+    try {
+      deleteOldVersionHint(fs, versionHintFile(), nextVersion);
+      versionCommitSuccess = commitNewVersion(fs, tempMetadataFile, finalMetadataFile, nextVersion);
+      if (!versionCommitSuccess) {
+        // Users should clean up orphaned files after job fail.
+        // This may be too heavy. But it can stay that way for now.
+        String msg =
```

Comment:
Typos in error message, underlying exception is swallowed here

Reply:
If versionCommitSuccess=false,I think we should throw an exception because the commit never actually succeeded. If we swallow the exception here, the user experience will be very strange.

Reply:
When I say "original exception is swallowed", I mean that a different exception was thrown in "commitNewVersion" but we don't get to see that exception here. Instead we get this generic message.

Reply:
Inside commitNewVersion we throw all exceptions, and we throw this exception directly to the commit method. It can be caught. 
```
 catch (CommitStateUnknownException | CommitFailedException e) {
      this.shouldRefresh = e instanceof CommitStateUnknownException || versionCommitSuccess;
      throw e;
    }
```


When versionCommitSuccess=false, it means that the rename operation failed because the dst exists, and fs.rename will not throw the exception. So I threw the exception manually.